lib/nrf_modem_lib/nrf_modem_lib_trace.c

```diff
@@ -27,9 +39,94 @@ static char rtt_buffer[CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_RTT_BUF_SIZE];
 
 static bool is_transport_initialized;
 
+struct trace_data_t {
+	void *fifo_reserved; /* 1st word reserved for use by fifo */
+	const uint8_t * const data;
+	uint32_t len;
+};
+
+K_FIFO_DEFINE(trace_fifo);
+
+#define TRACE_THREAD_STACK_SIZE 256
+#define TRACE_THREAD_PRIORITY CONFIG_NRF_MODEM_LIB_TRACE_THREAD_PRIO
+
+void trace_handler_thread(void)
+{
+	while (1) {
+		struct trace_data_t *trace_data = k_fifo_get(&trace_fifo, K_FOREVER);
+		const uint8_t * const data = trace_data->data;
+		const uint32_t len = trace_data->len;
+
 #ifdef CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_UART
+		/* Split RAM buffer into smaller chunks to be transferred using DMA. */
+		const uint32_t MAX_BUF_LEN = (1 << UARTE1_EASYDMA_MAXCNT_SIZE) - 1;
+		uint32_t remaining_bytes = len;
+		nrfx_err_t err;
+
+		while (remaining_bytes) {
+			size_t transfer_len = MIN(remaining_bytes, MAX_BUF_LEN);
+			uint32_t idx = len - remaining_bytes;
+
+			if (k_sem_take(&tx_sem, K_MSEC(UART_TX_WAIT_TIME_MS)) != 0) {
+				LOG_WRN("UARTE TX not available!");
+				break;
+			}
+			err = nrfx_uarte_tx(&uarte_inst, &data[idx], transfer_len);
+			if (err != NRFX_SUCCESS) {
+				LOG_ERR("nrfx_uarte_tx error: %d", err);
+				k_sem_give(&tx_sem);
+				break;
+			}
+			remaining_bytes -= transfer_len;
+		}
+		/* Wait for last UART transfer to finish */
+		k_sem_take(&tx_sem, K_FOREVER);
```

Comment:
Why do you take this semaphore twice, here and on line 70?
This thread should just block until the semaphore is available, waiting `K_FOREVER` in one place.

Reply:
Because we're splitting the buffer into chunks that fit the UARTE periperal and sending them one by one asynchronously. We need to wait for the UARTE callback to give the semaphore back before sending the next chunk. On line 83 we're waiting for the last chunk to be fully transferred before calling the `nrf_modem_trace_processed_callback`.

Reply:
To avoid this confusion, may be we should consider using `k_sem_count_get()`. What do you think ?

Reply:
Nevermind. @lemrey 's suggestion in https://github.com/nrfconnect/sdk-nrf/pull/6930#discussion_r823509665 is better than what I am trying to suggest.